[RFC] make unibi_out output never cross buffer boundaries

florolf opened the pull request:
TL;DR: cdfaecb caused a regression w.r.t. the problem fixed in #5436. This will hopefully fix this once and for all. See the commit messages for detailed reasoning.

florolf committed:
tui: never flush buffers in midst of unibilium output

e83845285 fixed an issue where long (true color) escape sequences got
interrupted by the cursor visibility toggling caused by buffer flushes.

cdfaecb25 introduces a new issue which causes similar problems: While
the old buffer flushing code appended the cursor visibility escapes to
the buffer before/after flushing, the new code effectively prepends the
sequences.

Assume the following sequence of events occurs:

 - A long escape code is issued using unibi_out when the buffer is
   almost full
 - out() gets called for a prefix of that escape code, causing the
   buffer to fill up
 - flush_buf(ui, false) is called and (correctly) does not insert any
   cursor toggling escapes
 - The rest of the escape code is written into the now empty buffer
 - At some later point, some other part of nvim calls flush_buf(ui,
   true), which then toggles the cursor, corrupting the escape code

This could possibly also be fixed by tracking the state of the buffer
(i.e. does it contain a partially output escape code?), but this seems
fragile in the same way e83845285 turned out to be.

The root cause for all these problems is the mismatch between nvim's
(implicit) assumption that the buffer is flushable at any point in time
and the non-atomicity of unibilium's character based callback interface.
The proper fix (without modifying unibilium) is to ensure nvim's
assumption about the buffer state holds at all times.

To that end, add a "cork" flag which ensures one unibi_out-call never
splits its output across a buffer flush; if an escape code does not fit
into the current buffer, flush it without any part of the escape code in
it and insert the whole escape code in the emptied buffer. This is a
little more complex because it modifies the buffer in place rather than
printing into another buffer, checking the remaining space in the
terminal buffer and then memcpy'ing it.

justinmk commented:
Greatly appreciated (again) :) 

see also https://github.com/neovim/neovim/issues/7381#issuecomment-348246065

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -1240,11 +1260,31 @@ static void out(void *ctx, const char *str, size_t len)
+  size_t available = sizeof(data->buf) - data->bufpos;
+
+  if (out_ctx->overflow)
+    return;
```
linter won't like this, i'll change it during the merge if CI passes.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -1240,11 +1260,31 @@ static void out(void *ctx, const char *str, size_t len)
+  data->bufpos += len;
+}
+
+static void limited_out(void *ctx, const char *str, size_t len)
```
Any reason not to replace `out()` with this entirely? And maybe store the state on `TUIData` instead of the new `limited_out_ctx` struct.

florolf replied:
I didn't want to modify the current users of `out()`, both because this slightly changes the semantics (possibly causing flushes at different points) and requires them to pass a context struct.

The semantics part shouldn't be an issue though, so one could reimplement `out()` using `limited_out()` which would reduce duplicated code. I'll do that.

Rolling `llimited_out_ctx` into `TUIData` feels slightly wrong because this is local state, but I have no strong opinion on this. If you think this is better stored as part of `TUIData`, I'll move it there. your call.

justinmk replied:
> The semantics part shouldn't be an issue though, so one could reimplement out() using limited_out() which would reduce duplicated code. I'll do that.

Ok, will wait for that before merging.

> Rolling llimited_out_ctx into TUIData feels slightly wrong because this is local state, but I have no strong opinion on this. If you think this is better stored as part of TUIData, I'll move it there. Your call.

Let's leave it as you have it. TUIData has fields like `bufpos`, but what you have is probably clearer and otherwise harmless.

florolf committed:
tui: always hide cursor when flushing

The previous commit ensures that we can never flush the buffer in a
state where toggling cursor visibility can corrupt other escape codes.
Thus, we can remove the workaround added as part of e838452, simplyfing
the code and hiding the cursor on more occasions.

florolf commented:
Only got to looking into this again today, sorry.

This is a new approach: The patches now introduce a "cork" state which prevents the buffer from being flushed (in cases where it ordinarily would be flushed, an overflow flag is raised and no further output is done). This reduces the code duplication somewhat.

When looking at the other users of `out` I realized that `tui_set_title` is another candidate for corruption by visibility toggling, but fixing this with the current infrastructure is a little more difficult. Once could leave the "do not toggle"-option of `flush_buf` in and use that in `tui_set_title`, but that would make things more complex. We could also simply truncate the title, as a 64k char long title probably doesn't make much sense. Either way, this probably is an issue for another PR.

florolf commented:
Sorry for the push noise, I had a long day. Should be okay now.
